### DISCUSSION TITLE
Questions to developers of ACore about custom versions of ACore

### DISCUSSION THREAD

**Discussion by Sulan2** [azerothcore/azerothcore-wotlk#10750] [General]
Hey all I wanted to ask the developers of acore what their thoughts are on people developing custom versions of ACore because this is a nightmare to update if u are making your own unique server
Im scared everytime I have to update in case I lose all my database adjustments but the reason I want to is because of being compatable with the latest version of Keira3 because that program is so damn good
With this said, Keira3 in my opinion is a program made to enhance custom game changes and also made by what I belive are people who made ACore so it doesnt make sense that the 2 doesnt function perfectly together

Is custom unique versions of ACore something u care about or is the goal of ACore only to make 1 perfect WOTLK version?

I got a server with a lot of custom changes made thru Keira3 that I started working on about half a year ago

I recompiled to a new version today (My old version was from november) and the only way to make worldserver start was to do everything manually, when u search online for help the answer always is having a value of 7 in worldserver config file at enable auto database updater but for people with a lot of database changes this doesnt work

The changes I made to make worldserver run was adding database tables that didnt exist and add new columns/rearrange old columns but this was incredibly rough to do

If I could wish for something for future I would prefer if the Vital SQL files would be seperated and noted out from the minor changes  SQL files by putting them in different folders
Basicaly u have things that is needed to make worldserver run (new tables that didnt exist before, new columns or rearranged columns in current tables) in 1 folder
(None of these changes causes problems on custom servers)

And then u have a folder with SQL files containing actual number changes inside the tables/columns
(These are problem files)

**Comment by Kitzunu** [comment]
The point of AC is to make it blizzlike.

And custom stuff is absolutely possible if you do it right.
You have modules and an entire folder for custom SQLs where you store your custom SQL changes so they do not get lost https://www.azerothcore.org/wiki/sql-directory

**Reply by Sulan2** [reply]
Putting SQLs in Custom folder apparently isnt as easy as putting an entire table in their which means I would have to spend a month time to backtrack all changes ive done in Keira and rewrite a SQL to work in custom folder?
If this is the best way to do this why doesnt Keira3 by itself write the files necessery for custom folder server updating?
Or is there any program that automatical writes the files needed?

Just seperating new SQLs based on what they do sounds a lot easier to me at least and will make ACore more compatible or all

**Reply by Nyeriah** [reply]
Keira3 gives you the SQL lines that you can paste into your backup files. Unfortunately, there's not much that can be done in regards to your previous work if you never made any backups of them.

**Reply by Sulan2** [reply]
Sometimes i do 300 changes in 1 day, im supposed to sit and copy paste and make SQL files of everything?

**Reply by Nyeriah** [reply]
Well... yes? 😄 If you want to ensure your work is never lost, yes, you should backup what you do.

**Reply by Kitzunu** [reply]
If you want to keep your custom stuff, yes. But it is up to what you want

**Reply by Sulan2** [reply]
What about providing an efficent and enjoyable work flow?
Again, all that is needed is new SQLs to be seperated and u would never have to copy paste
I dont know how often you add/change table and column structure but It doesnt feel like it happens too often
The majority of SQL updates are raw number changes?

Between my old version from November to now it was only about 30 changes

**Reply by Kitzunu** [reply]
The way AC handles SQLs will not change as we strive to be blizzlike.
We merge over 100 SQL fixes a month at this point and the system works fine.
in the automatic updater all SQL's in the Custom folder is applied last, aka your custom changes will always be applied last.

Maybe you can add a suggestion to Keira to output SQL files into a desired folder :)

**Reply by Sulan2** [reply]
All u need to do to is write 1 text file when u push the next the Version of ACore that says which new SQL file contains the database structure changes
I assume the amount of adjustments is like 5 or lower in each new version? since it was 30 from november 2021 to now

or entire ACore community needs to sit and copy paste every single line they want to adjust

Which one feels most efficent?

**Reply by Nyeriah** [reply]
The problem you describe is virtually a problem every developer has to face regardless of what they do or where they work. Backups need to be made to ensure data doesn't get lost, that has nothing to do with AzerothCore in particular, it's part of our daily lives.

It's not a major problem to us because the vital content we work with is preserved by commiting it to the repository, while the custom things we do are also shared inside modules to the OpenSource community.

I would suggest opening a suggestion at the Keira3 repository as Kitzunu suggested. There's not much we can do to help you other than what has already been said.

**Reply by Sulan2** [reply]
The way I see it is that its a new way to work with this server that u dont see yet that far ourweighs the currenty version for custom server workers and wouldnt impatct the current work flow of ACore

When I first made my server half a year ago I have in every new version I recompile compltely removed all SQLs files in the update folder aswell and therefor still havent lost any of my changes Ive ever made yet
This however is why I belive the worldserver auto db update 7 feature doesnt not work for me because the worldserver cannot start by itself
My way of working all works thanks to Trial and error working with the file DBerrors.log who has told me how to structure the database to work with the latest ACore version
Big credits to who ever created this log file

I dont think im asking for much, all I would want is for someone to write a few lines on what table has had its columns renamed to rearranged or new tables added etc (Basicaly exactly what the DBError.log file does)
Like here is a new version of ACore and here are just some short info about structure changes in the databases

I dont know how to track these changes because im not good at github but u guys are and because it feels like its in reality so few changes I doesnt feel like it would be more than 10mins of work to completely revolutionize the work flow in ACore for everyone who doesnt work the meta way

**Reply by Sulan2** [reply]
I can start this off if u want to get a feel for what I mean

Here is everything I had to do in terms of acore_world database structure changes from
AzerothCore rev. 86329a43ab8f 2021-10-29
to 
AzerothCore rev. c9586a106b4b+ 2022-02-19

----
In Creature table rename column called ID to ID1
Make 2 new columns called ID2 and ID3 with MEDIUMINT type and 0 default value
The new columns should be positioned as column 3 and 4

Run the following SQL files located in source acore-wotlk -> data -> sql -> Base
(Only spell.dbc and broadcast_text was tables I already had but its not modified anything anyway so no fear that none of my personal changes would be overwritted)
creature_template_movement.sql
creature_movement_override sql
playercreateinfo_cast_spell.sql
player_loot_template.sql
Broadcast_text.sql
Skilltiers.dbc
Spell.dbc

In Waypoint table add 2 columns
Orientation FLOAT type and 0 default value as column 6
delay INT type and 0 default value as column 7

In Creature_template table
Add 2 columns
speed_swim FLOAT type and 1 default value as Column nr 22
speed_flight Float type and 1 default value as Column nr 23

Remove Inhabit type column

In spell_proc_event table add a new column called procPhase
(For some reason this isnt in the base original files but worldserver.exe called for it)
I have no idea what they value is supposed to be here, I just set default value to 0

In Creature_template_addon and Creature_addon tables
Change name of "isLarge" column to "visibilityDistanceType"

In Npc_trainer table add column called ReqSpell INT Type and 0 default value as Column 7

----

The way I see it is that everything u have created in Acore_world since then until now that is NOT in this list are optional changes for people who want to play the Blizzlike ACore you want to create.
If u want to create a custom version of ACore like me the above "kind of info" is all the info u need
